File: ios/App/App/ScreenCaptureGuardPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * Fires a JS "captureDetected" event when the user takes a screenshot
 * (UIApplication.userDidTakeScreenshotNotification) or starts screen
 * recording/mirroring (UIScreen.capturedDidChangeNotification, checked
 * against UIScreen.main.isCaptured) while a call is active. Started/stopped
 * around active calls only (src/lib/screen-capture-guard.ts) via
 * startWatching/stopWatching, not for the app's whole lifetime.
 *
 * Android has no equivalent for "recording" — there's no public API there
 * for detecting screen recording by another app, only a screenshot callback
 * (API 34+, see ScreenCaptureGuardPlugin.java) — so "recording" is an
 * iOS-only capture kind.
 */
@objc(ScreenCaptureGuardPlugin)
public class ScreenCaptureGuardPlugin: CAPPlugin {
    private var watching = false

    @objc func startWatching(_ call: CAPPluginCall) {
        if !watching {
            watching = true
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(handleScreenshot),
                name: UIApplication.userDidTakeScreenshotNotification,
                object: nil
            )
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(handleScreenCaptureChanged),
                name: UIScreen.capturedDidChangeNotification,
                object: nil
            )
        }
        call.resolve()
    }

    @objc func stopWatching(_ call: CAPPluginCall) {
        if watching {
            watching = false
            NotificationCenter.default.removeObserver(self, name: UIApplication.userDidTakeScreenshotNotification, object: nil)
            NotificationCenter.default.removeObserver(self, name: UIScreen.capturedDidChangeNotification, object: nil)
        }
        call.resolve()
    }

    @objc private func handleScreenshot() {
        notifyListeners("captureDetected", data: ["kind": "screenshot"])
    }

    @objc private func handleScreenCaptureChanged() {
        // capturedDidChangeNotification fires on both the start AND end of
        // recording/mirroring — only the transition into isCaptured==true is
        // worth alerting on, the other direction is just "it stopped".
        if UIScreen.main.isCaptured {
            notifyListeners("captureDetected", data: ["kind": "recording"])
        }
    }
}

File: android/app/src/main/java/com/yukon3t/app/ScreenCaptureGuardPlugin.java
package com.yukon3t.app;

import android.app.Activity;
import android.os.Build;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Fires a JS "captureDetected" event when the user takes a screenshot while
 * this app is foregrounded, via Activity.registerScreenCaptureCallback
 * (Android 14 / API 34+ only — silently does nothing below that, there's no
 * older public API for this). Unlike iOS (see ScreenCaptureGuardPlugin.swift),
 * Android has no public API at all for detecting screen *recording* by
 * another app, so this plugin only ever reports kind "screenshot".
 * Started/stopped around active calls only (src/lib/screen-capture-guard.ts),
 * not for the app's whole lifetime.
 */
@CapacitorPlugin(name = "ScreenCaptureGuard")
public class ScreenCaptureGuardPlugin extends Plugin {

    private Activity.ScreenCaptureCallback screenCaptureCallback;

    @PluginMethod
    public void startWatching(PluginCall call) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                Activity activity = getActivity();
                if (activity != null && screenCaptureCallback == null) {
                    screenCaptureCallback = () -> {
                        JSObject data = new JSObject();
                        data.put("kind", "screenshot");
                        notifyListeners("captureDetected", data);
                    };
                    activity.registerScreenCaptureCallback(
                        ContextCompat.getMainExecutor(getContext()),
                        screenCaptureCallback
                    );
                }
            }
        } catch (Throwable t) {
            // Best-effort — a call already in progress shouldn't fail over this.
        }
        call.resolve();
    }

    @PluginMethod
    public void stopWatching(PluginCall call) {
        try {
            Activity activity = getActivity();
            if (activity != null && screenCaptureCallback != null) {
                activity.unregisterScreenCaptureCallback(screenCaptureCallback);
            }
        } catch (Throwable t) {
            // Best-effort.
        }
        screenCaptureCallback = null;
        call.resolve();
    }
}
